python/vineyard/contrib/ml/tensorflow.py

```diff
@@ -16,4 +16,136 @@
+        meta['label_type_'] = i[1].numpy().dtype.name
+        meta['data_type_meta_'] = i[0].numpy().dtype.str
+        meta['label_type_meta_'] = i[1].numpy().dtype.str
+    return client.create_metadata(meta)
+
+def tf_dataframe_builder(client, value, builder, **kw):
+    meta = ObjectMeta()
+    meta['typename'] = 'vineyard::DataFrame'
+    print(len(value))
+    for feat, labels in value.take(1):
+        cols = list(feat.keys())
+    cols.append('target')
+    meta['columns_'] = to_json(cols)
+    for i in range(len(cols)):
+        ls = []
+        for feat, labels in value.take(len(value)):
+            if cols[i] == 'target':
+                ls.append(labels.numpy())
+            else:
+                ls.append(feat[cols[i]].numpy())
+        meta['__values_-key-%d' % i] = to_json(cols[i])
+        meta.add_member('__values_-value-%d' % i, builder.run(client, ls))
+    meta['__values_-size'] = len(cols)
+    meta['partition_index_row_'] = kw.get('partition_index', [0, 0])[0]
+    meta['partition_index_column_'] = kw.get('partition_index', [0, 0])[1]
+    meta['row_batch_index_'] = kw.get('row_batch_index', 0)
+    print(meta)
+    return client.create_metadata(meta)
+
+def tf_builder(client, value, builder, **kw):
+    for x,y in value.take(1):
+        if type(x) is dict:
+            return tf_dataframe_builder(client, value, builder, **kw)
+        else:
+            return tf_tensor_builder(client, value, **kw)
+
+def tf_tensor_resolver(obj):
+    meta = obj.meta
+    num = from_json(meta['num'])
+    data_shape = from_json(meta['data_shape_'])
+    label_shape = from_json(meta['label_shape_'])
+    data_name = meta['data_type_']
+    label_name = meta['label_type_']
+    data_type = normalize_dtype(data_name, meta.get('value_type_meta_', None))
+    label_type = normalize_dtype(label_name, meta.get('value_type_meta_', None))
+    data = np.frombuffer(memoryview(obj.member('buffer_data_')), dtype=data_type).reshape(data_shape)
+    label = np.frombuffer(memoryview(obj.member('buffer_label_')), dtype=label_type).reshape(label_shape)
+    data = tf.data.Dataset.from_tensor_slices((data, label))
+    return data
+
+def tf_dataframe_resolver(obj, resolver):
+    meta = obj.meta
+    columns = from_json(meta['columns_'])
+    if not columns:
+        return pd.DataFrame()
+    blocks = []
+    index_size = 0
+    for idx, name in enumerate(columns):
+        np_value = resolver.run(obj.member('__values_-value-%d' % idx))
+        index_size = len(np_value)
+        if BlockPlacement:
+            placement = BlockPlacement(slice(idx, idx + 1, 1))
+        else:
+            placement = slice(idx, idx + 1, 1)
+        values = np.expand_dims(np_value, 0)
+        blocks.append(Block(values, placement, ndim=2))
+    if 'index_' in meta:
+        index = resolver.run(obj.member('index_'))
+    else:
+        index = np.arange(index_size)
+    df = pd.DataFrame(BlockManager(blocks, [pd.Index(columns), index]))
+    df = df.copy()
```

Comment:
Why do you need a `copy()` here?

Reply:
It was recommended, I can delete that, it works.